lib/cdo/redshift_import.rb

```diff
@@ -0,0 +1,170 @@
+require 'cdo/redshift'
+require 'pg'
+
+class RedshiftImport
+  # Database Migration Service Replication Tasks load data from Aurora into staging Redshift tables with a prefix.
+  TEMP_TABLE_PREFIX = '_import_'.freeze
+  BACKUP_TABLE_PREFIX = '_old_'.freeze
+
+  # DMS Replication Tasks import data from the production MySQL database into temporary / staging tables prefixed with
+  # "_import_" to avoid disrupting usage of the Analytics/Reporting solutions during the lengthy Full Load process.
+  # Complete import of staged data by renaming the current table to "_old_", renaming the staging table, and then
+  # dropping the old table.
+  # @param schemas [String] Array of Redshift schemas in which to complete data import.
+  def self.complete_table_import(schemas)
+    schemas.each do |schema|
+      temporary_import_tables(schema).each do |import_table|
+        target_table = import_table.partition(TEMP_TABLE_PREFIX).last
+        backup_table = BACKUP_TABLE_PREFIX + target_table
+
+        CDO.log.info "Dropping existing table #{schema}.#{target_table} and renaming newly imported #{import_table}."
+
+        # Rename existing table to back it up, if it exists.
+        begin
+          rename_table(schema, target_table, backup_table)
+        rescue PG::UndefinedTable => undefined_table_error
+          # When a new table created in the source MySQL database is imported for the first time, there won't be an
+          # existing table in Redshift to backup.
+          CDO.log.info "Unable to back up table #{schema}.#{target_table} because it does not exist. #{undefined_table_error}"
+        end
+
+        # Make staging table the production table.
+        rename_table(schema, import_table, target_table)
+
+        # Rename the primary key from '_import_[foo]_primary' to '[foo]_primary' so that the next time the
+        # DMS Replication Task runs and creates the staging table it can create the primary key without being blocked
+        # by an existing primary key with that name.
+        # Also rename the backup table's primary key.
+        old_primary_key = RedshiftImport.primary_key(schema, backup_table)
+        if old_primary_key
+          rename_primary_key(
+            schema,
+            backup_table,
+            old_primary_key[:name],
+            BACKUP_TABLE_PREFIX + old_primary_key[:name],
+            old_primary_key[:columns]
+          )
+        end
+
+        primary_key = RedshiftImport.primary_key(schema, target_table)
+        if primary_key
+          rename_primary_key(
+            schema,
+            target_table,
+            primary_key[:name],
+            primary_key[:name].partition(TEMP_TABLE_PREFIX).last,
+            primary_key[:columns]
+          )
+        end
```

Comment:
I wonder if this logic should be moved into `rename_table` (if it's necessary for the table-rename to be correct), and maybe even de-duplicate some of the doubled logic between these two calls to `rename_primary_key`.